File: Java/src/jakopec/pomocno/PartijaAdapter.java
package jakopec.pomocno;

import com.google.gson.*;
import jakopec.model.Partija;

import java.lang.reflect.Type;

public class PartijaAdapter implements JsonSerializer<Partija>, JsonDeserializer<Partija> {
    @Override
    public JsonElement serialize(Partija src, Type typeOfSrc, JsonSerializationContext context) {
       return null;
    }

    @Override
    public Partija deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
            throws JsonParseException {
        JsonObject jsonObject = json.getAsJsonObject();
        int brojIgraca = jsonObject.get("igraci").getAsJsonArray().size();
        try {
        switch (brojIgraca){
            case 2:
                return context.deserialize(json, Class.forName("jakopec.model.PartijaDvaIgraca"));
            case 3:
                return context.deserialize(json, Class.forName("jakopec.model.PartijaTriIgraca"));
            case 4:
                return context.deserialize(json, Class.forName("jakopec.model.PartijaDvaPara"));
            default:
                return null;
        }
        } catch (ClassNotFoundException cnfe) {
            throw new JsonParseException("Unknown element type: ", cnfe);
        }
    }
}

File: Kotlin/src/main/kotlin/pomocno/PartijaAdapter.kt
package pomocno

import com.google.gson.*
import model.Partija
import java.lang.reflect.Type

/**
 * @author Matija Skrletović
 **/

class PartijaAdapter : JsonSerializer<Partija?>, JsonDeserializer<Partija?> {
    override fun serialize(
        src: Partija?,
        typeOfSrc: Type,
        context: JsonSerializationContext
    ): JsonElement? {
        return null
    }

    @Throws(JsonParseException::class)
    override fun deserialize(
        json: JsonElement,
        typeOfT: Type,
        context: JsonDeserializationContext
    ): Partija? {
        val jsonObject = json.asJsonObject
        val brojIgraca = jsonObject["igraci"].asJsonArray.size()
        return try {
            when (brojIgraca) {
                2 -> context.deserialize(json, Class.forName("PartijaDvaIgraca"))
                3 -> context.deserialize(json, Class.forName("PartijaTriIgraca"))
                4 -> context.deserialize(json, Class.forName("PartijaDvaPara"))
                else -> null
            }
        } catch (cnfe: ClassNotFoundException) {
            throw JsonParseException("Unknown element type: ", cnfe)
        }
    }
}